packages/material-ui/src/FormControl/FormControl.js

```diff
@@ -49,114 +48,121 @@ export const styles = {
  *
  * ⚠️ Only one input can be used within a FormControl.
  */
-class FormControl extends React.Component {
-  static getDerivedStateFromProps(props, state) {
-    if (props.disabled && state.focused) {
-      return { focused: false };
-    }
-    return null;
-  }
+const FormControl = React.forwardRef(function FormControl(props, ref) {
+  const {
+    children,
+    classes,
+    className,
+    component: Component,
+    disabled,
+    error,
+    fullWidth,
+    margin,
+    required,
+    variant,
+    ...other
+  } = props;
+  const [adornedStart] = React.useState(() => {
+    // We need to iterate through the children and find the Input in order
+    // to fully support server-side rendering.
+    let initialAdornedStart = false;
 
-  constructor(props) {
-    super();
+    if (children) {
+      React.Children.forEach(children, child => {
+        if (!isMuiElement(child, ['Input', 'Select'])) {
+          return;
+        }
+
+        const input = isMuiElement(child, ['Select']) ? child.props.input : child;
 
-    this.state = {
-      adornedStart: false,
-      filled: false,
-      focused: false,
-    };
+        if (input && isAdornedStart(input.props)) {
+          initialAdornedStart = true;
+        }
+      });
+    }
+    return initialAdornedStart;
+  });
 
+  const [filled, setFilled] = React.useState(() => {
     // We need to iterate through the children and find the Input in order
     // to fully support server-side rendering.
-    const { children } = props;
+    let initialFilled = false;
+
     if (children) {
       React.Children.forEach(children, child => {
         if (!isMuiElement(child, ['Input', 'Select'])) {
           return;
         }
 
         if (isFilled(child.props, true)) {
-          this.state.filled = true;
-        }
-
-        const input = isMuiElement(child, ['Select']) ? child.props.input : child;
-
-        if (input && isAdornedStart(input.props)) {
-          this.state.adornedStart = true;
+          initialFilled = true;
         }
       });
     }
+
+    return initialFilled;
+  });
+
+  const [focused, setFocused] = React.useState(false);
+
+  if (disabled && focused) {
```

Comment:
Wow, at first I was like: wtf is this. But then, looking twice at the logic, it looks valid 👌.

Reply:
This is the function component version of 
```js
static getDerivedStateFromProps(props, state) {
  if (props.disabled && state.focused) {
    return { focused: false };
  }
  return null;
}
```

[Hooks FAQ: How do I implement getDerivedStateFromProps?](https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops)

Reply:
Thanks for the link.